### PR TITLE
Added support to Parse ASYNC function

### PR DESCRIPTION
**Pull request by mvvsmk**
Fixes issue #2650 
The parser now parses ASYNC functions.
gcc/rust/ChangeLog:

	* parse/rust-parse-impl.h (Parser::parse_item):Added ASYNC case.
	(Parser::parse_vis_item):Added a switch case to handle ASYNC.

gcc/testsuite/ChangeLog:

	* rust/compile/issue-2650.rs: New test.

Signed-off-by : M V V S Manoj Kumar <mvvsmanojkumar@gmail.com>

### CODE REVIEW

**Comment by dkm** [comment]
Typo in commit log (handle).
Also, you probably want to refer to #2650 instead of #2560

**Comment by mvvsmk** [comment]
Thanks for pointing out the typo and the link error. I have reworded my commit to reflect it.

File: gcc/rust/parse/rust-parse-impl.h
Comment (CohenArthur) [inline gcc/rust/parse/rust-parse-impl.h:1408]: async is not allowed if the Rust edition is set to 2015, which is the default. I think keeping the parser like that is fine, but we should add that to the checks in the AST validation pass

File: gcc/testsuite/rust/compile/issue-2650.rs
Comment (CohenArthur) [inline gcc/testsuite/rust/compile/issue-2650.rs:3]: so this test should eventually fail, unless we are in an edition >= 2018. so please do the following:
```suggestion
// { dg-additional-options "-frust-edition=2018" }

pub async fn a() -> u32 {
    1
}
```

**Review by CohenArthur — APPROVED** [review]

**Comment by CohenArthur** [comment]
example of the error if the Rust edition is set to 2015: https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=9696f6f21bd504217a3b9dab4947f507

File: gcc/rust/parse/rust-parse-impl.h
Comment (P-E-P) [reply gcc/rust/parse/rust-parse-impl.h:1408]: Rustc error out on `async` during parsing. We probably want to do the same to ensure macro compatibility.

File: gcc/rust/parse/rust-parse-impl.h
Comment (P-E-P) [inline gcc/rust/parse/rust-parse-impl.h:1390]: I feel like we could have a new function here, something along `parse_async_item`, we already have `parse_const_item`.

**Review by P-E-P — APPROVED** [review]
I'll be happy to merge this once the edition thing will be sorted out.

File: gcc/rust/parse/rust-parse-impl.h
Comment (mvvsmk) [reply gcc/rust/parse/rust-parse-impl.h:1408]: Could you tell me where or how I can find which version of Rust is set ? I know the magic happens in the lang.opt file, but how do we know what parameters are passed?

File: gcc/testsuite/rust/compile/issue-2650.rs
Comment (mvvsmk) [reply gcc/testsuite/rust/compile/issue-2650.rs:3]: I see, I'll do that. I am not familiar with dejagnu , but wouldn't it be "-frust-edition>=2018" or does it handle it on its own?

File: gcc/rust/parse/rust-parse-impl.h
Comment (mvvsmk) [reply gcc/rust/parse/rust-parse-impl.h:1390]: Understood, I'll make this change.

File: gcc/testsuite/rust/compile/issue-2650.rs
Comment (P-E-P) [reply gcc/testsuite/rust/compile/issue-2650.rs:3]: This dejagnu directive pass the argument to the compiler, this means the test code will be compiled using gccrs with the flag `-frust-edition=2018` hence why we're not setting `>=`, because we're setting the exact rust edition, we're not checking against all editions greater than 2018.

Comment (mvvsmk) [reply gcc/testsuite/rust/compile/issue-2650.rs:3]: Oh , I got it now, thanks for the help :D

File: gcc/rust/parse/rust-parse-impl.h
Comment (P-E-P) [reply gcc/rust/parse/rust-parse-impl.h:1408]: AFAIK there is a singleton in the rust session manager.

File: gcc/rust/parse/rust-parse-impl.h
Comment (mvvsmk) [reply gcc/rust/parse/rust-parse-impl.h:1390]: Hey @P-E-P , in our parser we have a function called `parse_async_block`, which has just has a defination and no caller in the entire codebase, would you like me to make a new `parse_async_item` function or just ammend my code to `parse_async_block`

Comment (P-E-P) [reply gcc/rust/parse/rust-parse-impl.h:1390]: I think `parse_async_block` was made exclusively for async block (`async {}`) you should probably create a new function instead of using this one.

Comment (mvvsmk) [reply gcc/rust/parse/rust-parse-impl.h:1390]: Oh ok, got it :)

**Comment by mvvsmk** [comment]
I have made the requested changes, if I missed anything do let me know :). 

I also added another test that fails, so we have both cases covered. I also noticed that our `parse_async_block` has no such check, would you like me to add it in this PR itself or make a separate one?

also a minor change but in rustc error says :
```
`async fn` is not permitted in Rust 2015
```
but the cpp11 compiler just kept giving me warnings about the " ` " symbol don't know what that was about, spent half the day figuring it out but it would still give me a warning no matter what I did so, changed it to :
```
'async fn' is not permitted in Rust 2015
```
if you guys got any hints about it do let me know :)

**Comment by P-E-P** [comment]
> I have made the requested changes, if I missed anything do let me know :).
> 
> I also added another test that fails, so we have both cases covered. I also noticed that our `parse_async_block` has no such check, would you like me to add it in this PR itself or make a separate one?
> 
> also a minor change but in rustc error says :
> 
> ```
> `async fn` is not permitted in Rust 2015
> ```
> 
> but the cpp11 compiler just kept giving me warnings about the " ` " symbol don't know what that was about, spent half the day figuring it out but it would still give me a warning no matter what I did so, changed it to :
> 
> ```
> 'async fn' is not permitted in Rust 2015
> ```
> 
> if you guys got any hints about it do let me know :)

That's because of quoted string. If the quoted thing lives in a variable you might use `%qX` with X being the correct format argument. For `*char` that would be `%qs`. Otherwise you can use `%<` and `%>`.

File: gcc/rust/parse/rust-parse-impl.h
Comment (P-E-P) [inline gcc/rust/parse/rust-parse-impl.h:1447]: `%<` already puts a ` in your output text, you do not need to keep those `

File: gcc/testsuite/rust/compile/issue-2650-2.rs
Comment (P-E-P) [inline gcc/testsuite/rust/compile/issue-2650-2.rs:3]: You won't be able to match the backtick token using dejaGNU because AFAIK it is based on the locale. User a dot `.` to match any character instead.

Comment (mvvsmk) [reply gcc/testsuite/rust/compile/issue-2650-2.rs:3]: Sure, I'll get it done.

File: gcc/rust/parse/rust-parse-impl.h
Comment (mvvsmk) [reply gcc/rust/parse/rust-parse-impl.h]: Done :D

**Review by CohenArthur — APPROVED** [review]
LGTM :) nice work!

File: gcc/rust/parse/rust-parse-impl.h
Comment (CohenArthur) [reply gcc/rust/parse/rust-parse-impl.h:1463]: once this is addressed we should merge the PR

Comment (mvvsmk) [reply gcc/rust/parse/rust-parse-impl.h]: I have made this change but I was experimeting a bit with rustc and found this error code 
```
error: expected one of `extern`, `fn`, or `unsafe`, found keyword `const`
 --> src/lib.rs:1:7
  |
1 | async const fn foo() {
  | ------^^^^^
  | |     |
  | |     expected one of `extern`, `fn`, or `unsafe`
  | help: `const` must come before `async`: `const async`
  ```
  would you like me to go with that and make the relavent changes by adding extern to the next token list and change the error message to the one emmited by rustc?
  
  this is the code I was testing : https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=38fc87fda8f9debc717354d3af2ef863

Comment (P-E-P) [reply gcc/rust/parse/rust-parse-impl.h]: @mvvsmk I think it would fit better in another PR

**Comment by mvvsmk** [comment]
@P-E-P Could you add this back to the merge queue when you get time, there was a merge conflict so it was removed from the queue.

**Pull request merged**
